feat: STREAMP-12186: adding disable flags for entity status support

andykruth-ii opened the pull request:
Adding flags to disable Entity Status.

* Implemented an old TODO to separate repository save functions into two that will allow specification and status saving to be done separately. This will help in disabling entity status support
* Updated all Entity Service Classes to use these distinct save methods
* Updated all Entity Mutation Implementations to make the updateStatus mutation a NOOP which does not persist the status when a toggle is disabled
* Added a convenience constructor to all Entity Model classes that does not require a status value.
* Updated the DefaultEntityView, DefaultEntityViewUpdater and KafkaEventReceiver with option disable toggle for entity status

andykruth-ii commented on core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ConsumerBindingService.java:
```diff
@@ -1,5 +1,5 @@
 /**
```
All entity service classes are updated to use separated `saveSpecification` and `saveStatus` repository methods.

andykruth-ii commented on graphql/api/src/main/java/com/expediagroup/streamplatform/streamregistry/graphql/mutation/impl/ConsumerBindingMutationImpl.java:
```diff
@@ -77,7 +80,12 @@ public Boolean delete(ConsumerBindingKeyInput key) {
+      return consumerBindingService.updateStatus(consumerBinding, status.asStatus()).get();
+    } else {
+      return consumerBinding;
+    }
```
All of the individual entity mutation implementation classes have this new Spring value toggle, set to true by default which is the current mode of operation.

When set to false an updateStatus mutation will NOT update the entity status via the entity service, it will just be a successful NOOP.

andykruth-ii commented on graphql/api/src/test/java/com/expediagroup/streamplatform/streamregistry/graphql/InputHelper.java:
```diff
@@ -0,0 +1,28 @@
+  public static StatusInput statusInput() {
+    return StatusInput.builder().agentStatus(mapper.createObjectNode()).build();
+  }
+}
```
Just a convenience test class to generate a statusInput

andykruth-ii commented on graphql/api/src/test/java/com/expediagroup/streamplatform/streamregistry/graphql/mutation/impl/ConsumerBindingMutationImplTest.java:
```diff
@@ -100,6 +101,39 @@ public void deleteWithCheckExistDisabledWhenEntityDoesNotExist() {
+    assertEquals(consumerBinding.get(), result);
+  }
+
   private ConsumerBindingKeyInput getConsumerBindingInputKey() {
```
Each entity has these two additional tests to test that when entity status is enabled (default) the entity service is called to update the status. The second test is for when entity status is disabled, the entity service should NOT be called when an updateStatus mutation is received.

andykruth-ii commented on repository/api/src/main/java/com/expediagroup/streamplatform/streamregistry/repository/Repository.java:
```diff
@@ -20,8 +20,16 @@
+
   Optional<T> findById(ID id);
 
   List<T> findAll();
```
Finally handling a TODO from long ago, deprecating the `save` method in favor of two separate `saveSpecification` and `saveStatus` methods.

andykruth-ii commented on state/core/src/main/java/com/expediagroup/streamplatform/streamregistry/state/DefaultEntityView.java:
```diff
@@ -47,10 +47,19 @@ public class DefaultEntityView implements EntityView {
+
   @Override
   public CompletableFuture<Void> load(@NonNull EntityViewListener listener) {
     val future = new CompletableFuture<Void>();
```
Adding an optional `entityStatusEnabled` flag to all constructor patterns will can be used to disable entity status persistence.

jamespfaulkner commented on repository/kafka/src/main/java/com/expediagroup/streamplatform/streamregistry/repository/kafka/DefaultRepository.java:
```diff
@@ -48,9 +48,19 @@ abstract class DefaultRepository<
+  public ME saveSpecification(ME entity) {
     Optional<ME> existing = findById(entity.getKey());
     Entity<SK, SS> stateEntity = converter.convertEntity(entity);
     List<CompletableFuture<Void>> futures = new ArrayList<>();
```
Do we need these lists of futures now? Could we do similar to what we do in the agents...

```java
  private void sendSync(Event<SK, SS> event) {
    sender.send(event).join();
  }
```

Usages then become:
```java
sendSync(Event.specification(stateEntity.getKey(), stateEntity.getSpecification()));
```

andykruth-ii replied:
yep, missed that. done

jamespfaulkner reviewed (commented):
Couple of questions:

Rather than having multiple instances of the `entityStatusEnabled ` in all instances of the `*MutationImpl` classes, would it be simpler to just ignore these in the `KafkaEventSender`? `enableStatuses` could be added to `Config` there, the sending could then be changed to something like:

```java 
  @Override
  public <K extends Entity.Key<S>, S extends Specification> CompletableFuture<Void> send(@NonNull Event<K, S> event) {
    if (!config.enableStatuses && (event instanceof StatusEvent<K, S> || event instanceof StatusDeletionEvent<K, S>)) {
      return CompletableFuture.completedFuture(null);
    }

    val avroEvent = converter.toAvro(event);
    return send(avroEvent.getKey(), avroEvent.getValue());
  }
```

We'd then remove the if/else in all `*MutationImpl`.

--------

Second question, and maybe this is progress for a follow up. Would it be possible to change the `Event` class hierarchy in allow easier differentiation between entity/status events. 

I'm thinking two new interfaces with `SpecificationEvent`/`StatusEvent`/`SpecificationDeletionEvent`/`StatusDeletionEvent` extending the correct interface.
```java 
public interface EntityEvent<K extends Entity.Key<S>, S extends Specification> extends Event<K, S> { }

public interface StatusEvent<K extends Entity.Key<S>, S extends Specification> extends Event<K, S> { }
```

We could then have a new `EventReceiverListener`  which deals only with the entity events (meaning agents really are shielded from Status events...

```java
<K extends Entity.Key<S>, S extends Specification> void onEvent(EntityEvent<K, S> event);
```

andykruth-ii commented on repository/kafka/src/main/java/com/expediagroup/streamplatform/streamregistry/repository/kafka/DefaultRepository.java:
```diff
@@ -48,9 +48,19 @@ abstract class DefaultRepository<
+   * {link {@link #saveStatus(com.expediagroup.streamplatform.streamregistry.model.Entity)}}
+   * to save the entity specification or entity status independently.
+   */
+  @Deprecated
```
yep good call, done

andykruth-ii commented:
@jamespfaulkner - OK summary of updates:
* Added the entity status toggle to KafkaEventSender similar to KafkaEventReceiver
* Better tests for KafkaEventReceiver
* Removed all MutationImpl changes, no longer necessary
* Removed all EntityView changes, no longer necessary
* removed repository.send method instead of deprecating so that we get a compile error instead of runtime error
* removed futures on single event send in saveSpecification

jamespfaulkner commented on state/kafka-receiver/src/main/java/com/expediagroup/streamplatform/streamregistry/state/kafka/KafkaEventReceiver.java:
```diff
@@ -38,12 +38,8 @@
-import lombok.NonNull;
-import lombok.RequiredArgsConstructor;
-import lombok.Value;
+import lombok.*;
```
Minor: star import
